File: leetcode/6-push-dominoes/myHan.java
class Solution {
    public String pushDominoes(String dominoes) {
        char[] ret = dominoes.toCharArray();
        
        int[] when = new int[dominoes.length()];
        int[] q = new int[dominoes.length()];
        int front = 0, back = 0;
        
        for(int i = 0; i < dominoes.length(); i++){
            if(dominoes.charAt(i) != '.') {
                q[back++] = i;
            }
        }
        
        while(front != back) {
            int cur = q[front++];
            int nxt = cur - 1;
            if(ret[cur] == 'R')
                nxt = cur + 1;
            if(nxt < 0 || ret.length <= nxt)
                continue;
            
            if(ret[nxt] == '.') {
                ret[nxt] = ret[cur];
                when[nxt] = when[cur] + 1;
                q[back++] = nxt;
            }
            else if(ret[nxt] != ret[cur] && when[nxt] == when[cur] + 1) {
                ret[cur - 1] = '.';
            }
        }
        
        return new String(ret);
    }
}

File: leetcode/16-find-peak-element/myHan.ex
defmodule Solution do
  @spec search(%{integer => integer}, integer, integer) :: integer
  def search(nums, front, back) when front + 1 < back do
    mid = div(front + back, 2)
    prv = Map.get(nums, mid-1)
    cur = Map.get(nums, mid)
    nxt = Map.get(nums, mid+1)
    {nfront, nback} = cond do
      prv < cur && cur < nxt -> {mid, back}
      nxt < cur && prv < cur -> {mid, mid+1}
      true -> {front, mid}
    end
    search(nums, nfront, nback)
  end

  def search(_, front, _), do: front

  @spec find_peak_element(nums :: [integer]) :: integer
  def find_peak_element(nums) do
    mnums = nums
      |> Enum.zip(0 .. (length(nums) - 1))
      |> Enum.reduce(%{}, fn {v, k}, acc -> Map.put_new(acc, k, v) end)

    search(mnums, 0, length(nums))
  end
end
